File: attributes-fabric/src/main/java/io/github/solusmods/eternalcore/attributes/fabric/EternalCoreAttributeRegisterImpl.java
package io.github.solusmods.eternalcore.attributes.fabric;

import dev.architectury.event.events.common.LifecycleEvent;
import io.github.solusmods.eternalcore.attributes.api.EternalCoreAttributes;
import net.fabricmc.fabric.api.object.builder.v1.entity.FabricDefaultAttributeRegistry;
import net.minecraft.core.Holder;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.ai.attributes.Attribute;
import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
import net.minecraft.world.entity.ai.attributes.DefaultAttributes;
import net.minecraft.world.entity.ai.attributes.RangedAttribute;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

@SuppressWarnings("unchecked_cast")
public class EternalCoreAttributeRegisterImpl {
    public static final List<Holder<Attribute>> GENERIC_REGISTRY = new CopyOnWriteArrayList<>();
    public static final List<Holder<Attribute>> PLAYER_REGISTRY = new CopyOnWriteArrayList<>();

    public static Holder<Attribute> registerToPlayers(Holder<Attribute> holder) {
        PLAYER_REGISTRY.add(holder);
        return holder;
    }

    public static Holder<Attribute> registerToGeneric(Holder<Attribute> holder) {
        GENERIC_REGISTRY.add(holder);
        return holder;
    }

    public static @NotNull Holder<Attribute> registerPlayerAttribute(String modID, String id, String name, double amount,
                                                                     double min, double max, boolean syncable, Attribute.Sentiment sentiment) {
        Attribute attribute = new RangedAttribute(name, amount, min, max).setSyncable(syncable).setSentiment(sentiment);
        Holder<Attribute> holder = Registry.registerForHolder(BuiltInRegistries.ATTRIBUTE, EternalCoreAttributes.getResourceKey(modID, id), attribute);
        return registerToPlayers(holder);
    }

    public static @NotNull Holder<Attribute> registerGenericAttribute(String modID, String id, String name, double amount,
                                                                      double min, double max, boolean syncable, Attribute.Sentiment sentiment) {
        Attribute attribute = new RangedAttribute(name, amount, min, max).setSyncable(syncable).setSentiment(sentiment);
        Holder<Attribute> holder = Registry.registerForHolder(BuiltInRegistries.ATTRIBUTE, EternalCoreAttributes.getResourceKey(modID, id), attribute);
        return registerToGeneric(holder);
    }

    public static AttributeSupplier.Builder addLivingEntityAttributes(AttributeSupplier.Builder builder) {
        for (Holder<Attribute> holder : GENERIC_REGISTRY) builder.add(holder);
        return builder;
    }

    @SuppressWarnings("unchecked_cast")
    public static void init() {
        LifecycleEvent.SETUP.register(() -> {
            BuiltInRegistries.ENTITY_TYPE.stream().filter(DefaultAttributes::hasSupplier)
                    .map(entityType -> (EntityType<? extends LivingEntity>) entityType)
                    .forEach(entityType -> {
                        if (entityType == null) return;

                        AttributeSupplier.Builder builder = new AttributeSupplier.Builder();
                        DefaultAttributes.getSupplier(entityType).instances.forEach((attribute, attributeInstance) -> {
                            builder.add(attribute, attributeInstance.getBaseValue());
                        });

                        GENERIC_REGISTRY.forEach(builder::add);
                        if (entityType.equals(EntityType.PLAYER)) PLAYER_REGISTRY.forEach(builder::add);
                        FabricDefaultAttributeRegistry.register(entityType, builder);
                    });
        });
    }
}

File: attributes-neoforge/src/main/kotlin/io/github/solusmods/eternalcore/attributes/neoforge/EternalCoreAttributeRegisterImpl.kt
package io.github.solusmods.eternalcore.attributes.neoforge

import net.minecraft.core.Holder
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.ai.attributes.Attribute
import net.minecraft.world.entity.ai.attributes.RangedAttribute
import net.neoforged.fml.ModLoadingContext
import net.neoforged.neoforge.event.entity.EntityAttributeModificationEvent
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer
import java.util.function.Supplier

object EternalCoreAttributeRegisterImpl {
    private val GENERIC_REGISTRY: MutableList<Holder<Attribute?>?> = CopyOnWriteArrayList<Holder<Attribute?>?>()
    private val PLAYER_REGISTRY: MutableList<Holder<Attribute?>?> = CopyOnWriteArrayList<Holder<Attribute?>?>()

    fun registerToPlayers(holder: Holder<Attribute?>): Holder<Attribute?> {
        PLAYER_REGISTRY.add(holder)
        return holder
    }

    fun registerToGeneric(holder: Holder<Attribute?>): Holder<Attribute?> {
        GENERIC_REGISTRY.add(holder)
        return holder
    }

    @JvmStatic
    fun registerPlayerAttribute(
        modID: String?, id: String, name: String, amount: Double,
        min: Double, max: Double, syncable: Boolean, sentiment: Attribute.Sentiment
    ): Holder<Attribute?> {
        val attribute = RangedAttribute(name, amount, min, max).setSyncable(syncable).setSentiment(sentiment)
        return registerToPlayers(
            EternalCoreAttributeNeoForge.attributes.register(
                id,
                Supplier { attribute })
        )
    }

    @JvmStatic
    fun registerGenericAttribute(
        modID: String?, id: String, name: String, amount: Double,
        min: Double, max: Double, syncable: Boolean, sentiment: Attribute.Sentiment
    ): Holder<Attribute?> {
        val attribute = RangedAttribute(name, amount, min, max).setSyncable(syncable).setSentiment(sentiment)
        return registerToGeneric(
            EternalCoreAttributeNeoForge.attributes.register<Attribute?>(
                id,
                Supplier { attribute })
        )
    }

    fun registerAttributes(e: EntityAttributeModificationEvent) {
        e.types.forEach(Consumer { type: EntityType<out LivingEntity?>? ->
            if (type == EntityType.PLAYER) PLAYER_REGISTRY.forEach(Consumer { holder: Holder<Attribute?>? ->
                e.add(
                    type,
                    holder
                )
            })
            GENERIC_REGISTRY.forEach(Consumer { holder: Holder<Attribute?>? -> e.add(type, holder) })
        })

        // Clear the registry
        PLAYER_REGISTRY.clear()
        GENERIC_REGISTRY.clear()
    }

    @JvmStatic
    fun init() {
        val modEventBus = ModLoadingContext.get().activeContainer.eventBus
        if (modEventBus == null) return
        modEventBus.addListener { obj: EntityAttributeModificationEvent? -> registerAttributes(obj!!) }
    }
}
